rasa/core/domain.py

```diff
@@ -690,103 +673,107 @@ def input_states(self) -> List[Text]:
+    ) -> Dict[Text, Dict[Text, Union[Text, Tuple[Text]]]]:
         state_dict = {}
 
         # Set all found entities with the state value 1.0, unless they should
         # be ignored for the current intent
         latest_message = tracker.latest_message
 
-        if not latest_message:
+        if not latest_message or latest_message == UserUttered.empty():
             return state_dict
 
-        intent_name = latest_message.intent.get(INTENT_NAME_KEY)
-
-        if intent_name:
-            for entity_name in self._get_featurized_entities(latest_message):
-                key = f"entity_{entity_name}"
-                state_dict[key] = 1.0
+        state_dict[USER] = latest_message.as_dict_core()
 
-        # Set all set slots with the featurization of the stored value
-        for key, slot in tracker.slots.items():
-            if slot is not None:
-                if slot.value == "None" and slot.as_feature():
-                    # TODO: this is a hack to make a rule know
-                    #  that slot or form should not be set
-                    #  but only if the slot is featurized
-                    slot_id = f"slot_{key}_None"
-                    state_dict[slot_id] = 1
-                else:
-                    for i, slot_value in enumerate(slot.as_feature()):
-                        if slot_value != 0:
-                            slot_id = f"slot_{key}_{i}"
-                            state_dict[slot_id] = slot_value
-
-        if "intent_ranking" in latest_message.parse_data:
-            for intent in latest_message.parse_data["intent_ranking"]:
-                if intent.get(INTENT_NAME_KEY):
-                    intent_id = "intent_{}".format(intent[INTENT_NAME_KEY])
-                    state_dict[intent_id] = intent["confidence"]
-
-        elif intent_name:
-            intent_id = "intent_{}".format(latest_message.intent[INTENT_NAME_KEY])
-            state_dict[intent_id] = latest_message.intent.get("confidence", 1.0)
+        # filter entities based on intent config
+        entities = tuple(self._get_featurized_entities(latest_message))
+        if entities:
+            state_dict[USER][ENTITIES] = entities
+        else:
+            del state_dict[USER][ENTITIES]
 
         return state_dict
 
     def _get_featurized_entities(self, latest_message: UserUttered) -> Set[Text]:
         intent_name = latest_message.intent.get(INTENT_NAME_KEY)
         intent_config = self.intent_config(intent_name)
         entities = latest_message.entities
-        entity_names = {
+        entity_names = set(
             entity["entity"] for entity in entities if "entity" in entity.keys()
-        }
+        )
 
         wanted_entities = set(intent_config.get(USED_ENTITIES_KEY, entity_names))
 
         return entity_names.intersection(wanted_entities)
 
-    def get_prev_action_states(
-        self, tracker: "DialogueStateTracker"
-    ) -> Dict[Text, float]:
+    @staticmethod
+    def _get_slots_states(
```

Comment:
I think this should rather just return the state dict and not a wraper (the `{SLOTS: ...}` part)

Reply:
it return empty dict, if state dict is empty. returning `{SLOTS: ...}` allows us to make this check inside this method, otherwise will need to check it outside